src/ripple/basics/PerfLog.h

```diff
@@ -0,0 +1,157 @@
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_PERFLOG_H
+#define RIPPLE_BASICS_PERFLOG_H
+
+#include <ripple/protocol/TER.h>
+#include <ripple/core/JobTypes.h>
+#include <ripple/json/json_value.h>
+#include <string>
+#include <cstdint>
+#include <memory>
+#include <type_traits>
+#include <chrono>
+
+namespace ripple {
+namespace perf {
+
+/**
+ * Singleton class that maintains performance counters and optionally
+ * writes Json-formatted data to a distinct log. It should exist prior
+ * to other objects launched by Application to make it accessible for
+ * performance logging.
+ */
+
+class PerfLog
+{
+public:
+    /**
+     * Configuration from [perf] section of rippled.cfg.
+     */
+    struct Setup
+    {
+        std::string perf_log;
+        std::uint64_t log_interval {1};
+    };
+
+    /**
+     * Log start of RPC call.
+     *
+     * @param method RPC command
+     * @param requestId Unique identifier to track command
+     */
+    virtual void rpcStart(std::string const& method,
+        std::uint64_t requestId) = 0;
+
+    /**
+     * Log successful finish of RPC call
+     *
+     * @param method RPC command
+     * @param requestId Unique identifier to track command
+     */
+    virtual void rpcFinish(std::string const& method,
+        std::uint64_t requestId) = 0;
+
+    /**
+     * Log errored RPC call
+     *
+     * @param method RPC command
+     * @param requestId Unique identifier to track command
+     */
+    virtual void rpcError(std::string const& method,
+        std::uint64_t requestId) = 0;
+
+    /**
+     * Log queued job
+     *
+     * @param type Job type
+     */
+    virtual void jobQueue(std::underlying_type_t<JobType> const type) = 0;
+
+    /**
+     * Log job executing
+     *
+     * @param type Job type
+     * @param dur Duration enqueued in microseconds
+     * @param startTime Time that execution began
+     * @param instance JobQueue worker thread instance
+     */
+    virtual void jobStart(std::underlying_type_t<JobType> const type,
+        std::uint64_t dur,
```

Comment:
Type-unsafe API here: if you _really_ want `dur` to be microseconds, then you ought to just make this `std::chrono::microseconds`.

Looking at the docs, a `std::chrono::microseconds` has at least 55 bits of resolution which works out (and @HowardHinnant can keep me honest here) to almost 23 centuries. That should be enough for our use.

Reply:
Technically 55 bits only gets us +/- 570 years.  But as there are no machines which have 55bit words, we get 64 bits or +/- 292 thousand years. :-)

If I ever get a chance to re-specify those ranges, the minimum is going to be +/- 32K years, except for nanoseconds which we can't afford more than +/- 292 years.  This formula would put microseconds at a minimum requirement of 61 bits.  Technically we can do it without any backwards compatibility concerns.  But I'd have to catch the committee either drop dead tired, or during a host-paid open bar. ;-)

I agree with Nik's suggestion, and this is an especially fun and safe change to make.  Just change `std::uint64_t dur` to `std::chrono::microseconds dur` right where Nik points to, and then start compiling.  You'll get compile-time errors.  Just do what the compiler tells you to in order to make that initial change compile.  It _might_ talk you into changing other durations stored as `uint64_t` into `std::chrono::microseconds`, or it might not (e.g. `struct Jq`).  You decide where to draw the line between the chrono API and integers.

But this change is safe in that it won't introduce run-time errors.  When you get it to compile, it will be right, wherever you decide to draw that line.